server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestGetSnapshotsAction.java

```diff
@@ -53,6 +55,10 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         getSnapshotsRequest.ignoreUnavailable(request.paramAsBoolean("ignore_unavailable", getSnapshotsRequest.ignoreUnavailable()));
         getSnapshotsRequest.verbose(request.paramAsBoolean("verbose", getSnapshotsRequest.verbose()));
         getSnapshotsRequest.masterNodeTimeout(request.paramAsTime("master_timeout", getSnapshotsRequest.masterNodeTimeout()));
-        return channel -> client.admin().cluster().getSnapshots(getSnapshotsRequest, new RestToXContentListener<>(channel));
+        return channel -> new RestCancellableNodeClient(client, request.getHttpChannel()).admin().cluster().getSnapshots(
+                getSnapshotsRequest,
+                // no need to fork here, this will be called on the generic pool anyway if its a large response for more than the
+                // currently running snapshots
+                new DispatchingRestToXContentListener<>(EsExecutors.DIRECT_EXECUTOR_SERVICE, channel, request));
```

Comment:
Yea that was because we didn't have that check in the general listener. Should we maybe make that move in a separate PR? Even though it seems like "nothing could go wrong" by doing that, might be better having it in isolation?

Reply:
Sure, sounds like a good plan.